x-pack/metricbeat/module/openai/usage/client.go

```diff
@@ -0,0 +1,66 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package usage
+
+import (
+	"net/http"
+	"time"
+
+	"golang.org/x/time/rate"
+
+	"github.com/elastic/elastic-agent-libs/logp"
+)
+
+// RLHTTPClient implements a rate-limited HTTP client that wraps the standard http.Client
+// with a rate limiter to control API request frequency.
+type RLHTTPClient struct {
+	client      *http.Client
+	logger      *logp.Logger
+	Ratelimiter *rate.Limiter
+}
+
+// Do executes an HTTP request while respecting rate limits.
+// It waits for rate limit token before proceeding with the request.
+// Returns the HTTP response and any error encountered.
+func (c *RLHTTPClient) Do(req *http.Request) (*http.Response, error) {
+	start := time.Now()
+
+	c.logger.Debug("Waiting for rate limit token")
+
+	err := c.Ratelimiter.Wait(req.Context())
+	if err != nil {
+		return nil, err
```

Comment:
Shall we log this error: 
        c.logger.Errorf("Failed to acquire rate limit token: %v", err)

Reply:
I'll return the error with some context; it will get logged at the caller of the function it is residing in.

Reply:
Done.